src/backend/columnar/columnar_customscan.c

```diff
@@ -821,12 +838,117 @@ static void
 ColumnarScan_ExplainCustomScan(CustomScanState *node, List *ancestors,
 							   ExplainState *es)
 {
-	TableScanDesc scanDesc = node->ss.ss_currentScanDesc;
+	ColumnarScanDesc columnarScanDesc = (ColumnarScanDesc) node->ss.ss_currentScanDesc;
 
-	if (scanDesc != NULL)
+	if (columnarScanDesc != NULL)
 	{
-		int64 chunkGroupsFiltered = ColumnarScanChunkGroupsFiltered(scanDesc);
+		int64 chunkGroupsFiltered = ColumnarScanChunkGroupsFiltered(columnarScanDesc);
 		ExplainPropertyInteger("Columnar Chunk Groups Removed by Filter", NULL,
 							   chunkGroupsFiltered, es);
 	}
+
+	ColumnarScanState *columnarScanState = (ColumnarScanState *) node;
+	const char *projectedColumnsStr =
+		ColumnarScanProjectedColumnsStr(columnarScanState, ancestors, es);
+	ExplainPropertyText("Columnar Projected Columns", projectedColumnsStr, es);
+}
+
+
+/*
+ * ColumnarScanProjectedColumnsStr generates projected column string for
+ * explain output.
+ */
+static const char *
+ColumnarScanProjectedColumnsStr(ColumnarScanState *columnarScanState, List *ancestors,
+								ExplainState *es)
+{
+	ScanState *scanState = &columnarScanState->custom_scanstate.ss;
+
+	List *neededVarList = ColumnarVarNeeded(columnarScanState);
+	if (list_length(neededVarList) == 0)
+	{
+		return "<columnar optimized out all columns>";
+	}
+
+#if PG_VERSION_NUM >= 130000
+	List *context =
+		set_deparse_context_plan(es->deparse_cxt, scanState->ps.plan, ancestors);
+#else
+	List *context =
+		set_deparse_context_planstate(es->deparse_cxt, (Node *) &scanState->ps,
+									  ancestors);
+#endif
+
+	bool useTableNamePrefix = false;
+	bool showImplicitCast = false;
+	return deparse_expression((Node *) neededVarList, context,
+							  useTableNamePrefix, showImplicitCast);
+}
+
+
+/*
+ * ColumnarVarNeeded returns a list of Var objects for the ones that are
+ * needed during columnar custom scan.
+ * Throws an error if finds a Var referencing to an attribute not supported
+ * by ColumnarScan.
+ */
+static List *
+ColumnarVarNeeded(ColumnarScanState *columnarScanState)
+{
+	ScanState *scanState = &columnarScanState->custom_scanstate.ss;
+
+	List *varList = NIL;
+
+	Bitmapset *neededAttrSet = ColumnarAttrNeeded(scanState);
+	int bmsMember = -1;
+	while ((bmsMember = bms_next_member(neededAttrSet, bmsMember)) >= 0)
+	{
+		/* neededAttrSet represents 0-indexed attribute numbers */
+		AttrNumber attrNumber = bmsMember + 1;
+
+		Relation columnarRelation = scanState->ss_currentRelation;
+		HeapTuple columnTuple =
+			SearchSysCacheAttNum(RelationGetRelid(columnarRelation), attrNumber);
+		if (!HeapTupleIsValid(columnTuple))
+		{
```

Comment:
You get get this information more easily with `columnarRelation->rd_att->attrs[attrNumber]`